pocketutils/misc/klgists.py

```diff
@@ -0,0 +1,829 @@
+	
+	p = subprocess.Popen(cmd, stdout=PIPE, stderr=PIPE, cwd=cwd, bufsize=bufsize)
+	try:
+		q = Queue()
+		Thread(target=_reader, args=[PipeType.STDOUT, p.stdout, q]).start()
+		Thread(target=_reader, args=[PipeType.STDERR, p.stderr, q]).start()
+		for _ in range(2):
+			for source, line in iter(q.get, None):
+				log_callback(source, line)
+		exit_code = p.wait(timeout=timeout_secs)
+	finally:
+		p.kill()
+	if exit_code != 0:
+		raise ResourceError("Got nonzero exit code {} from '{}'".format(exit_code, ' '.join(cmd)), cmd, exit_code, '<<unknown>>', '<<unknown>>')
+	
+
+def wrap_cmd_call(cmd: List[str], stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell_cmd: str=None, cwd: Optional[str] = None, timeout_secs: Optional[float] = None) -> (str, str):
+	"""Calls an external command, waits, and throws a ResourceError for nonzero exit codes.
+	Returns (stdout, stderr).
+	The user can optionally provide a shell to run the command with, e.g. "powershell.exe" 
+	"""
+	cmd = [str(p) for p in cmd]
+	if shell_cmd:
+		cmd = [shell_cmd] + cmd
+	logger.debug("Calling '{}'".format(' '.join(cmd)))
+	p = subprocess.Popen(cmd, stdout=stdout, stderr=stderr, cwd=cwd)
+	out, err, exit_code = None, None, None
+	try:
+		(out, err) = p.communicate(timeout=timeout_secs)
+		out = out.decode('utf-8')
+		err = err.decode('utf-8')
+		exit_code = p.wait(timeout=timeout_secs)
+	except Exception as e:
+		_log(out, err, logger.warning)
+		raise e
+	finally:
+		p.kill()
+	if exit_code != 0:
+		_log(out, err, logger.warning)
+		raise ResourceError("Got nonzero exit code {} from '{}'".format(exit_code, ' '.join(cmd)), cmd, exit_code, out, err)
+	_log(out, err, logger.debug)
+	return out, err
+
+def look(obj: object, attrs: str) -> any:
+	if not isinstance(attrs, str) and isinstance(attrs, Iterable): attrs = '.'.join(attrs)
+	try:
+		return operator.attrgetter(attrs)(obj)
+	except AttributeError: return None
+
+def flatmap(func, *iterable):
+	return chain.from_iterable(map(func, *iterable))
+
+def flatten(*iterable):
+	return list(chain.from_iterable(iterable))
+
+class DevNull:
+	def write(self, msg): pass
+
+pjoin = os.path.join
+pexists = os.path.exists
+pdir = os.path.isdir
+pfile = os.path.isfile
+pis_dir = os.path.isdir
+fsize = os.path.getsize
+def pardir(path: str, depth: int=1):
+	for _ in range(-1, depth):
+		path = os.path.dirname(path)
+	return path
+def grandpardir(path: str):
+	return pardir(path, 2)
+
+
+T = TypeVar('T')
+def try_index_of(element: List[T], list_element: T) -> Optional[T]:
+	try:
+		index_element = list_element.index(element)
+		return index_element
+	except ValueError:
+		return None
+
+def decorator(cls):
+	return cls
+
+
+def exists(keep_predicate: Callable[[T], bool], seq: Iterable[T]) -> bool:
+	"""Efficient existential quantifier for a filter() predicate.
+	Returns true iff keep_predicate is true for one or more elements."""
+	for e in seq:
+		if keep_predicate(e): return True  # short-circuit
+	return False
+
+
+def zip_strict(*args):
```

Comment:
This is `CoreTools.zip_strict`, but Python 10 added a [`strict=True` flag](https://peps.python.org/pep-0618/).